packages/network-controller/src/NetworkController.ts

```diff
@@ -696,130 +773,141 @@ export class NetworkController extends BaseControllerV2<
+   * Adds a new custom network or updates the information for an existing
+   * network.
+   *
+   * This may involve updating the `networkConfigurations` property in
+   * state as well and/or adding a new network client to the network client
+   * registry. The `rpcUrl` and `chainId` of the given object are used to
+   * determine which action to take:
+   *
+   * - If the `rpcUrl` corresponds to an existing network configuration
+   * (case-insensitively), then it is overwritten with the object. Furthermore,
+   * if the `chainId` is different from the existing network configuration, then
+   * the existing network client is replaced with a new one.
+   * - If the `rpcUrl` does not correspond to an existing network configuration
+   * (case-insensitively), then the object is used to add a new network
+   * configuration along with a new network client.
    *
-   * @param networkConfiguration - The network configuration to add or, if rpcUrl matches an existing entry, to modify.
-   * @param networkConfiguration.rpcUrl -  RPC provider url.
-   * @param networkConfiguration.chainId - Network ID as per EIP-155.
-   * @param networkConfiguration.ticker - Currency ticker.
-   * @param networkConfiguration.nickname - Personalized network name.
-   * @param networkConfiguration.rpcPrefs - Personalized preferences (i.e. preferred blockExplorer)
-   * @param options - additional configuration options.
-   * @param options.setActive - An option to set the newly added networkConfiguration as the active provider.
-   * @param options.referrer - The site from which the call originated, or 'metamask' for internal calls - used for event metrics.
-   * @param options.source - Where the upsertNetwork event originated (i.e. from a dapp or from the network form) - used for event metrics.
-   * @returns id for the added or updated network configuration
+   * @param networkConfiguration - The network configuration to add or update.
+   * @param options - Additional configuration options.
+   * @param options.referrer - Used to create a metrics event; the site from which the call originated, or 'metamask' for internal calls.
+   * @param options.source - Used to create a metrics event; where the event originated (i.e. from a dapp or from the network form).
+   * @param options.setActive - If true, switches to the network upon adding or updating it (default: false).
+   * @returns The ID for the added or updated network configuration.
    */
   async upsertNetworkConfiguration(
-    { rpcUrl, chainId, ticker, nickname, rpcPrefs }: NetworkConfiguration,
+    networkConfiguration: NetworkConfiguration,
     {
-      setActive = false,
       referrer,
       source,
-    }: { setActive?: boolean; referrer: string; source: string },
+      setActive = false,
+    }: {
+      referrer: string;
+      source: string;
+      setActive?: boolean;
+    },
   ): Promise<string> {
-    assertIsStrictHexString(chainId);
+    const sanitizedNetworkConfiguration = (
```

Comment:
We were technically doing this before by destructuring the object that this method takes, but that approach meant that properties of the network configuration object that were not technically provided ended up as `undefined` when put into `networkConfigurations`. The code here is essentially equivalent to Lodash's `pick` function.

Reply:
Makes sense! Potentially simpler to use the argument as-is and throw if extra properties are there (safer too, as we'd be informed about these properties being ignored rather than them being silently dropped). But that would be better as a separate PR, if we did want to go that route.